drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/rulesengine/SessionStatsCollector.java

```diff
@@ -117,7 +121,24 @@ public void registerMatch(RulesExecutorSession session, Match match) {
     }
 
     public void registerMatchedEvents(Collection<FactHandle> events) {
-        matchedEvents += events.size();
+        long count = events.stream()
+                .filter(this::isCountedAsMatchedEvent)
+                .count();
+        matchedEvents += (int) count;
+    }
+
+    private boolean isCountedAsMatchedEvent(FactHandle fh) {
+        Object obj = fh.getObject();
+        if (!(obj instanceof PrototypeFactInstance prototypeFactInstance)) {
+            // (Usually, not a case) Not a PrototypeFactInstance, so we count it as a matched event
+            return true;
+        }
+        if (!SYNTHETIC_PROTOTYPE_NAME.equals(prototypeFactInstance.getPrototype().getName())) {
+            // Not a synthetic event, so we count it as a matched event
+            return true;
+        }
+        // Unless a specific flag is set, we do not count a synthetic event as a matched event
+        return Boolean.TRUE.equals(prototypeFactInstance.get(COUNT_AS_MATCHED_EVENT));
     }
```

Comment:
Firstly, I fixed `registerMatchedEvents()` to exclude synthetic events (control events) to count. However, I noticed that  `OnceAfterDefinition` and `TimedOutDefinition` expect that the main rule is fired by a control event. So, I added a boolean field `count_as_matched_event` to flag if the control event should be counted for the SessionStats.